src/core/function-core.js

```diff
@@ -0,0 +1,115 @@
+const process = require('process');
+const fetch = require('node-fetch');
+
+const BASE_URL = process.env.FUNCTION_BASE_URL;
+
+function createChatForTwoUsers(matchRow) {
+  //console.log('createChatForTwoUsers');
+  const current_url = `${BASE_URL}/addNewChatBetweenUsers?userId1=${matchRow.userId1}&userId2=${matchRow.userId2}`;
+  //console.log(current_url);
+  return fetch(current_url, {
+    method: 'GET',
+    headers: {
+      'Content-Type': 'application/json',
+      'FUNCTION_SECRET_KEY': process.env.FUNCTION_SECRET_KEY
+    }
+  })
+  .then(res => res.json())
+  .then(data => {
+    return data.chatKey;
+  })
+};
+
+function addPushNotificationTokenForUserId(userId, pushToken) {
+  //console.log('addPushNotificationTokenForUserId');
+  const current_url = `${BASE_URL}/addPushTokenForUserId?userId=${userId}&pushToken=${pushToken}`;
+  //console.log(current_url);
+  return fetch(current_url, {
+    method: 'GET',
+    headers: {
+      'Content-Type': 'application/json',
+      'FUNCTION_SECRET_KEY': process.env.FUNCTION_SECRET_KEY
+    }
+  })
+  .catch(err => {
+    console.log(err);
+    console.log('something went wrong in addPushNotificationTokenForUserId')
```

Comment:
Fail strategy has not been implemented.

What would you suggest?

We can eg. try again in 15 seconds and if that fails, too, log something. Would that be good? Of course the assumption is that it should not fail but...

What would be a good yet simple thing to do here and on the other similar cases?

Reply:
fetch doesn't support retries out of the box, but libraries like superagent and got do have it. I think it's sensible to retry a connection once or twice in case of a network failure, but I would consider push notifications as a best effort kind of feature. Personally I'd leave it at a few retries

Reply:
I agree that simply retrying a reasonable number of times and (maybe) with escalating intervals is sufficient. Most likely user will receive more than a single notification during a conversations lifetime. If one notification is not delivered, it might be likely that he'll get another notification from a different message anyway. Think the worst case would be that neither party gets the initial notification indicating that a new conversation is now live, but that might be such a fringe case anyway that it is not worth solving at this time.

Reply:
At least [got](https://github.com/sindresorhus/got) does the retries with delays, calculated like this:

```javascript
1000 * Math.pow(2, retry) + Math.random() * 100
```

Just blindly hammering the retries one after another might lead to another quick set of failures, so a delay between retries makes sense

Reply:
@cc-akarvinen